### PR TITLE
feat: add map block to globals

### PR DESCRIPTION
**Pull request by i4ki**
# Reason for This Change

Introduces the `map` block inside the `globals` block.
The `map` block was designed to solve some problems of `{for ...}` comprehensions and `tm_merge()`, related to constructing complex objects out of lists and merging objects with duplicated keys.

The `map` block has the structure below:

```
map <var name> {
    for_each = <expr>    // expr must return a list
    iterator = <keyword> // defines the name of the iteration variable (default is `element`)
    key      = <expr>    // defines the key for the value
    value    = <expr>    // defines the value for the key.

    value {
        <attr> = <expr>
        ...
        
        map <var name> {
            // recursive...
        }
    }
}
```
The value must be defined by either a `value` attribute or a block, but never both.
When iterating the values defined in the `for_each` attribute, the variable defined by the iterator (by default it's called `element`) will contain the object below:

```hcl
element = {
    new = <current element>
    old = <previous set value in the object> || nil
}
```

When using the `value` block, it can contain child `map` blocks which in case allows the creation of nested objects.

See the usage below:

```hcl
globals {
  people = ["marius", "tiago", "tiago", "soeren"]
  map "people_count" {
    for_each = global.people

    key = element.new  # element.new is each name

    value {
      count = tm_try(element.old.count, 0) + 1
    }
  }
}
```
The code above will generate the object below:
```hcl
global.people_count = {
  marius = {count = 1}
  tiago  = {count = 2}
  soeren = {count = 1}
}
```

### CODE REVIEW

File: stack/globals_test.go [hunk 1]
```diff
@@ -2796,6 +2796,758 @@ func TestLoadGlobals(t *testing.T) {
...
+			},
+			want: map[string]*hclwrite.Block{
+				"/stack": Globals(
+					EvalExpr(t, "var", `{
+						something = "else"
+					}`),
+				),
+			},
+		},
+		{
```
Comment (katcipis) [inline stack/globals_test.go:2978]: maybe now would be a good time to split the tests into different files, like tests focused on map, tests focused on other global behavior, etc. The current test table is quite big. Not a huge thing, just wondering.

**Review by katcipis — APPROVED** [review]
LGTM + some comments that can be addressed later

File: mapexpr/map.go [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+			},
+		}
+	}
+
+	evaluator.DeleteNamespace(m.Attrs.Iterator)
+	return cty.ObjectVal(objmap), nil
+}
+
+// Variables returns the variables referenced by the map block.
+// TODO(i4k): implement.
```
[the remark this answers is not in the document]
Comment (i4ki) [reply mapexpr/map.go:220]: I'm gonna implement this in a separate PR.

File: stack/globals_test.go [hunk 3]
```diff
@@ -2796,6 +2796,758 @@ func TestLoadGlobals(t *testing.T) {
 			},
 		},
 		{
+			name:   "globals.map missing the label",
+			layout: []string{"s:stack"},
+			configs: []hclconfig{
+				{
+					path: "/",
+					add: Globals(
+						Map(),
```
[the remark this answers is not in the document]
Comment (i4ki) [reply stack/globals_test.go:2805]: ah yes, this was the first test I added while working in the block validation. Thanks

**Review by katcipis — APPROVED** [review]
:rocket: :rocket: :rocket:

**Pull request merged**